### ISSUE TITLE
Allow the host to make the option type non-nestable and erased

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fsharp/fslang-suggestions#546]
From @dsyme's comment [here](https://github.com/fable-compiler/fable-react/commit/e85554a61e797db8eb3202f6f35d49a4b1d0a9d0#commitcomment-21001608)

> BTW I think you should add an F# language suggestion to allow the "host" (i.e. Fable) to make the "option" type be non-nestable and erased, so 'T option would add a constraint that 'T is not an option type.

TBD

- How the user can tell the compiler and the IDE who's the host (.NET F# compiler or Fable). At the moment, Fable only identifies itself by defining the `FABLE_COMPILER` constant.
- Are there other checks that the compiler could do if it knows options will be erased? For example, forbidding things like `Map<'a, 'b option>` as it would cause problems with `Map.tryFind`; representing the result of empty Active Patterns with something else than `Some ()` (this is a special case in Fable, as the runtime representation of both `()` and `None` is null), etc.

**Estimated cost**: M-L 

## Affadavit (must be submitted)

* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

* [x] This is not a breaking change to the F# language design
* [x] I would be willing to help implement and/or test this
* [ ] I or my company would be willing to help crowdfund F# Software Foundation members to work on this

**Comment by dsyme** [comment]
> representing the result of empty Active Patterns with something else than Some () (this is a special case in Fable, as the runtime representation of both () and None is null), etc.

I suppose the constraint is ``doesn't use null as a true value``.  That constraint would also be useful more generally in F#

> •	Are there other checks that the compiler could do if it knows options will be erased? For example, forbidding things like Map<'a, 'b option> as it would cause problems with Map.tryFind; 

I suppose the second type parameter of ``Map`` would have the constraint.

**Comment by alfonsogarciacaro** [comment]
BTW, Fable 1.3 accepts nested options. It has now a mixed model of erased options and a `Some` runtime type to solve ambiguities (like nested options, unit options, etc). It may seem complicated but again it was the only way to solve some of the problems mentioned above, like optional arguments or representing TypeScript optional members.

**Comment by dsyme** [comment]
I'll close this since Fable no longer has a direct need for this.  It might make sense for some other transpilers

**Comment by alfonsogarciacaro** [comment]
I totally forgot about this! Funny enough I actually had the same problem recently when adding Dart compilation to Fable. Dart is a typed language and has null-safety, so types that can be null need to be marked explicitly. At first I tried to represent this explicit nullable types with optional types in F#, but this also prevented nested options and was not favored by other team members. In case anyone is interested, this is how I resolved null/option interoperability with Dart (at least for now): https://github.com/fable-compiler/Fable/issues/2919#issuecomment-1146480435
